File: apps/inference-service/app/emitter.py
import hashlib
import hmac
import json
import logging
from typing import Any

import httpx

from .config import settings

logger = logging.getLogger("inference.emitter")


class EventEmitter:
    """Posts proctor events to the backend's external webhook.

    Authenticates with an HMAC-SHA256 signature over the raw JSON body using the
    shared PROCTOR_WEBHOOK_SECRET — the same zero-trust contract the Electron
    agent and any external provider use. The backend risk engine applies
    weighting + decay; this service only forwards observations as evidence.
    """

    def __init__(self) -> None:
        self._client = httpx.AsyncClient(timeout=5.0)

    async def emit(
        self,
        session_id: str,
        event_type: str,
        severity: str = "MEDIUM",
        payload: dict[str, Any] | None = None,
    ) -> bool:
        body = json.dumps(
            {
                "sessionId": session_id,
                "type": event_type,
                "severity": severity,
                "payload": payload or {},
                "provider": "inference-service",
            },
            separators=(",", ":"),
        )
        signature = hmac.new(
            settings.proctor_webhook_secret.encode(),
            body.encode(),
            hashlib.sha256,
        ).hexdigest()

        try:
            res = await self._client.post(
                f"{settings.api_base_url}/api/proctoring/webhook",
                content=body,
                headers={
                    "content-type": "application/json",
                    "x-proctor-signature": f"sha256={signature}",
                },
            )
            return res.status_code < 300
        except httpx.HTTPError as exc:  # network hiccup — never crash the pipeline
            logger.warning("Failed to emit %s: %s", event_type, exc)
            return False

    async def aclose(self) -> None:
        await self._client.aclose()


emitter = EventEmitter()

File: apps/desktop-agent/src/emitter.ts
import { createHmac } from 'node:crypto';

/**
 * Signed event emitter.
 *
 * Posts proctoring events to the backend's external webhook, authenticating
 * with an HMAC-SHA256 signature over the raw JSON body (shared
 * PROCTOR_WEBHOOK_SECRET). This is the same trust model the backend expects
 * from any external proctor provider — no user JWT required.
 */
export interface AgentConfig {
  apiBaseUrl: string;   // e.g. https://app.interviewai.example.com
  webhookSecret: string;
  sessionId: string;
}

export interface AgentEvent {
  type: string;
  severity?: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  payload?: Record<string, unknown>;
}

export class EventEmitterClient {
  constructor(private readonly cfg: AgentConfig) {}

  async emit(event: AgentEvent): Promise<boolean> {
    const body = JSON.stringify({
      sessionId: this.cfg.sessionId,
      type: event.type,
      severity: event.severity ?? 'MEDIUM',
      payload: event.payload ?? {},
      provider: 'desktop-agent',
    });
    const signature = createHmac('sha256', this.cfg.webhookSecret).update(body).digest('hex');

    try {
      const res = await fetch(`${this.cfg.apiBaseUrl}/api/proctoring/webhook`, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          'x-proctor-signature': `sha256=${signature}`,
        },
        body,
      });
      return res.ok;
    } catch {
      return false;
    }
  }
}
